src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1475 @@
+            }
+
+            // Determines if the current class could be extended in some external compilation unit
+            private boolean currentClassIsExternallyExtendable() {
+                return !this.currentClass.sym.isFinal() &&
+                  !(this.currentClass.sym.isSealed() && this.currentClass.permitting.isEmpty()) &&
+                  !(this.currentClass.sym.owner.kind == MTH) &&
+                  !this.privateOuter;
+            }
+        }.scan(tree);
+
+        // TODO: eliminate sealed classes where all permitted subclasses are in this compilation unit
+
+        // Now analyze all of the analyzable constructors we found
+        for (Map.Entry<Symbol, MethodInfo> entry : this.methodMap.entrySet()) {
+
+            // We are looking for analyzable constructors only
+            final Symbol sym = entry.getKey();
+            final MethodInfo methodInfo = entry.getValue();
+            if (!methodInfo.isAnalyzable())
+                continue;
+
+            // Analyze constructor body
+            this.targetClass = methodInfo.getDeclaringClass();
+            this.methodClass = this.targetClass;
+            Assert.check(this.depth == -1);
+            Assert.check(this.refs == null);
+            this.pushScope();
+            try {
+
+                // Add the initial 'this' reference
+                this.refs = RefSet.newEmpty();
+                this.refs.add(ThisRef.direct());
+
+                // Scan constructor statements
+                this.analyzeStatements(methodInfo.getDeclaration().body.stats);
+            } finally {
+                this.popScope();
+                this.methodClass = null;
+                this.targetClass = null;
+                this.refs = null;
+            }
+        }
+
+        // Eliminate duplicate warnings. Warning B duplicates warning A if the stack trace of A is a prefix
+        // of the stack trace of B. For example, if constructor Foo(int x) has a leak, and constructor
+        // Foo() invokes this(0), then emitting a warning for Foo() would be redundant.
+        final BiPredicate<DiagnosticPosition[], DiagnosticPosition[]> extendsAsPrefix = (warning1, warning2) -> {
+            if (warning2.length < warning1.length)
+                return false;
+            for (int index = 0; index < warning1.length; index++) {
+                if (warning2[index].getPreferredPosition() != warning1[index].getPreferredPosition())
+                    return false;
+            }
+            return true;
+        };
+
+        // Stack traces are ordered top to bottom, and so duplicates always have the same first element(s).
+        // Sort the stack traces lexicographically, so that duplicates immediately follow what they duplicate.
+        final Comparator<DiagnosticPosition[]> ordering = (warning1, warning2) -> {
+            for (int index1 = 0, index2 = 0; true; index1++, index2++) {
+                final boolean end1 = index1 >= warning1.length;
```

Comment:
> From a documentation perspective it might carry a bit of value

Yes, that's the purpose - the `final` is for the human viewer, not the compiler. Just trying to be helpful to the next person.

But you're right it's inconsistent so I'll remove.